### PR TITLE
Add input limit check to prevent secp256k1-zkp panic when >256 inputs

### PR DESCRIPTION
**Pull request by tosynthegeek**
This PR addresses issue #95 where attempting to create a transaction with more than 256 inputs causes a panic due to secp256k1-zkp's internal limit on the number of surjection proof inputs.

Specifically:
- The limit `SECP256K1_SURJECTIONPROOF_MAX_N_INPUTS` (256) is now enforced across transaction creation logic.

- Checks are inserted before every possible call to add_input or input addition logic in `finish` and `finish_liquidex_take`

If the limit is reached, a `TooManyInputs` error is returned instead of panicking.

### CODE REVIEW

File: lwk_wollet/src/error.rs [hunk 1]
```diff
@@ -227,6 +227,9 @@ pub enum Error {
 
     #[error("Issuance amount greater than 21M*10^8 are not allowed")]
     IssuanceAmountGreaterThanBtcMax,
+
+    #[error("Transaction input {0} exceeds maximum allowed input count of 256")]
```
Comment (delta1) [inline lwk_wollet/src/error.rs:231]:
```suggestion
    #[error("Number of transaction inputs ({0}) exceeds maximum allowed input count of 256")]
```

File: lwk_wollet/src/tx_builder.rs [hunk 2]
```diff
@@ -799,13 +816,22 @@ impl TxBuilder {
...
                     wollet.add_input(&mut pset, &mut inp_txout_sec, &mut inp_weight, utxo)?;
                     satoshi_in += utxo.unblinded.value;
                 }
             }
             None => {
                 // FIXME: For implementation simplicity now we always add all L-BTC inputs
                 for utxo in wollet.asset_utxos(&wollet.policy_asset())? {
+                    if pset.inputs().len() >= SECP256K1_SURJECTIONPROOF_MAX_N_INPUTS {
+                        return Err(Error::TooManyInputs(pset.inputs().len()));
+                    }
```
Comment (delta1) [inline lwk_wollet/src/tx_builder.rs:833]: If we're doing this check every time before `add_input` then I think it would be better to just have this check inside that method instead. That way it's not duplicated in many places, and can't be forgotten if any new calls to `add_input` are added somewhere else.

**Review by delta1 — COMMENTED** [review]
Thanks for picking this up @tosynthegeek. 

Left some suggestions. Could you please also add a unit test for this?

**Comment by tosynthegeek** [comment]
Sure, would write unit tests for this

File: lwk_wollet/src/error.rs (hunk 1 above)
Comment (tosynthegeek) [reply lwk_wollet/src/error.rs]: Updated the attribute as suggested

File: lwk_wollet/src/tx_builder.rs (hunk 2 above)
Comment (tosynthegeek) [reply lwk_wollet/src/tx_builder.rs]: Makes sense. Updated the check to be done within the `add_input` method and also added unit tests for verifying the behavior when the input count reaches the maximum allowed limit as well as when it is just below the limit.

File: lwk_wollet/src/pset_create.rs [hunk 3]
```diff
@@ -245,4 +254,58 @@ mod test {
...
+
+        match result {
+            Err(Error::TooManyInputs(count)) => {
+                assert_eq!(count, SECP256K1_SURJECTIONPROOF_MAX_N_INPUTS);
+            }
+            _ => panic!("Expected TooManyInputs error"),
+        }
+    }
+
+    #[test]
```
Comment (delta1) [inline lwk_wollet/src/pset_create.rs:280]: I think this could all be in one test. 

Add 256 inputs in a loop (as you are already doing), asserting that each one is added Ok. 

Then after the loop add the 257th input and assert on the Err

File: lwk_wollet/src/pset_create.rs [hunk 4]
```diff
@@ -245,4 +254,58 @@ mod test {
...
+
+        for _ in 0..SECP256K1_SURJECTIONPROOF_MAX_N_INPUTS - 1 {
+            pset.add_input(Default::default());
+        }
+
+        let result = wollet.add_input(&mut pset, &mut inp_txout_sec, &mut inp_weight, &dummy_utxo);
+        assert!(result.is_ok());
+    }
+
+    // duplicated from tests/test_wollet.rs
```
[the remark this answers is not in the document]
Comment (tosynthegeek) [reply lwk_wollet/src/pset_create.rs:296]: It is located within the tests/ folder so I cant directly import it except I have to create a shared file

File: lwk_wollet/src/tx_builder.rs [hunk 5]
```diff
@@ -786,6 +786,7 @@ impl TxBuilder {
             if utxo.unblinded.asset != policy_asset {
                 continue;
             }
+
```
[the remark this answers is not in the document]
Comment (tosynthegeek) [reply lwk_wollet/src/tx_builder.rs]: Fixed

File: lwk_wollet/src/pset_create.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (delta1) [reply lwk_wollet/src/pset_create.rs:280]: Oh right sorry, the test should go along with the other tests and import from the public api

[the remark this answers is not in the document]
Comment (tosynthegeek) [reply lwk_wollet/src/pset_create.rs:280]: @delta1 I am expecting your feedback on this ser

[the remark this answers is not in the document]
Comment (RCasatta) [reply lwk_wollet/src/pset_create.rs:280]: It's not great but I think it's fine to duplicate this test code, other solution are suboptimal too

**Comment by tosynthegeek** [comment]
> Concept ACK [f43c844](https://github.com/Blockstream/lwk/commit/f43c844d145ba936e4ada2832aea4da698af6bd2)
> 
> Could you please squash the commits in two commits since there has been some back and forth? One with the code change and the other with the test

Done, can you please check

File: lwk_wollet/tests/e2e.rs [hunk 6]
```diff
@@ -2490,6 +2490,8 @@ fn test_no_wildcard_with_path_after() {
 }
 
 #[test]
+<<<<<<< master
```
Comment (LeoComandini) [inline lwk_wollet/tests/e2e.rs:2493]: please remove this

**Comment by LeoComandini** [comment]
@tosynthegeek it seems the rebase/squash has a bunch of extra commits

**Comment by tosynthegeek** [comment]
> @tosynthegeek it seems the rebase/squash has a bunch of extra commits

@LeoComandini there were some conflicts which resulted in extra commits

**Comment by LeoComandini** [comment]
@tosynthegeek we need something like this https://github.com/Blockstream/lwk/commits/tony-pr101/

If you want to change this PR to match that branch, we can wait.
But if you're fine with it, we can merge the branch linked above

**Comment by tosynthegeek** [comment]
> @tosynthegeek we need something like this https://github.com/Blockstream/lwk/commits/tony-pr101/
> 
> If you want to change this PR to match that branch, we can wait. But if you're fine with it, we can merge the branch linked above



> @tosynthegeek we need something like this https://github.com/Blockstream/lwk/commits/tony-pr101/
> 
> If you want to change this PR to match that branch, we can wait. But if you're fine with it, we can merge the branch linked above

I have resolved, please confirm now

**Review by RCasatta — APPROVED** [review]
ACK 20f2f1777e5388bd7f6a136b6e2f817204b97020

Thanks for the contribution and for the squash
